In [1]:
import pandas as pd
import numpy as np

Dataset Loading

In [2]:
df = pd.read_csv('/content/drive/MyDrive/DVA_2/Electric_Vehicle_Population_Data.csv')

/tmp/ipykernel_10118/1250250578.py:1: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/drive/MyDrive/DVA_2/Electric_Vehicle_Population_Data.csv')


### 2. Data Cleaning & Standardisation

#### Fill missing geographic data with 'Unknown'

In [3]:
geo_cols = ['County', 'City', 'Postal Code', 'Legislative District', 'Vehicle Location']
for col in geo_cols:
    df[col] = df[col].fillna('Unknown')
print("Missing geographic data filled with 'Unknown'.")

Missing geographic data filled with 'Unknown'.


#### Standardise 'Postal Code' as a string (remove '.0')

In [4]:
df['Postal Code'] = df['Postal Code'].astype(str).str.replace('.0', '', regex=False)
print("'Postal Code' standardised.")

'Postal Code' standardised.


### 3. Handling 'Electric Range'

#### Create 'Electric Range Raw' and treat 0 values in 'Electric Range' as NaN for statistical analysis

In [5]:
df['Electric Range Raw'] = df['Electric Range']
df['Electric Range'] = df['Electric Range'].replace(0, np.nan)
print("'Electric Range' handled: raw column created and 0 values replaced with NaN.")

'Electric Range' handled: raw column created and 0 values replaced with NaN.


### 4. Feature Engineering (KPI Prep)

#### Create a 'Vehicle Age' column

In [6]:
current_year = 2026
df['Vehicle Age'] = current_year - df['Model Year']
print("'Vehicle Age' column created.")

'Vehicle Age' column created.


#### Simplify 'CAFV Eligibility' for visualization

In [7]:
df['CAFV_Status'] = df['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].apply(
    lambda x: 'Eligible' if 'Eligible' in str(x) else 'Unknown/Not Eligible'
)
print("'CAFV_Status' column created.")

'CAFV_Status' column created.


### 5. Feature Engineering: Create 'Range_Category'

In [8]:
print("Intermediate export skipped. Final export will occur after all feature engineering.")

Intermediate export skipped. Final export will occur after all feature engineering.


### Displaying the first 5 rows of the cleaned data

In [9]:
display(df.head())

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,...,DOL Vehicle ID,Vehicle Location,Electric Utility,2020 Census Tract,Electric Range Raw,Vehicle Age,CAFV_Status,Range_Category,Latitude,Longitude
0,1N4AZ0CP6D,King,Kirkland,WA,98034,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,154635729,POINT (-122.22901 47.72201),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303302e+10,75.0,13,Eligible,Urban Commuter,47.72201,-122.22901
1,5YJ3E1EC8L,Kitsap,Bainbridge Island,WA,98110,2020,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,107518645,POINT (-122.521 47.62759),PUGET SOUND ENERGY INC,5.303509e+10,308.0,6,Eligible,Long Range,47.62759,-122.52100
2,5YJ3E1EBXJ,King,Seattle,WA,98144,2018,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,474808813,POINT (-122.30866 47.57874),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10,215.0,8,Eligible,Regional Traveler,47.57874,-122.30866
3,ZFAFFAC45R,Thurston,Yelm,WA,98597,2024,FIAT,500E,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,...,273658514,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10,NaN,2,Unknown/Not Eligible,Unknown/Not Researched,46.94239,-122.60735
4,5YJYGDEE3L,King,Kent,WA,98030,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,109579900,POINT (-122.19975 47.37483),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303303e+10,291.0,6,Eligible,Long Range,47.37483,-122.19975


### 6. Advanced Feature Engineering (for Tableau Visualization)

In [ ]:
bins = [-1, 0, 100, 250, 1000]
labels = ['Unknown/Not Researched', 'Urban Commuter', 'Regional Traveler', 'Long Range']

# Create the new categorical column
df['Range_Category'] = pd.cut(df['Electric Range Raw'], bins=bins, labels=labels)

print("'Range_Category' column created.")
display(df.head())

'Range_Category' column created.


,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,...,DOL Vehicle ID,Vehicle Location,Electric Utility,2020 Census Tract,Electric Range Raw,Vehicle Age,CAFV_Status,Range_Category,Latitude,Longitude
0,1N4AZ0CP6D,King,Kirkland,WA,98034,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,154635729,POINT (-122.22901 47.72201),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303302e+10,75.0,13,Eligible,Urban Commuter,47.72201,-122.22901
1,5YJ3E1EC8L,Kitsap,Bainbridge Island,WA,98110,2020,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,107518645,POINT (-122.521 47.62759),PUGET SOUND ENERGY INC,5.303509e+10,308.0,6,Eligible,Long Range,47.62759,-122.52100
2,5YJ3E1EBXJ,King,Seattle,WA,98144,2018,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,474808813,POINT (-122.30866 47.57874),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10,215.0,8,Eligible,Regional Traveler,47.57874,-122.30866
3,ZFAFFAC45R,Thurston,Yelm,WA,98597,2024,FIAT,500E,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,...,273658514,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10,NaN,2,Unknown/Not Eligible,NaN,46.94239,-122.60735
4,5YJYGDEE3L,King,Kent,WA,98030,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,109579900,POINT (-122.19975 47.37483),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303303e+10,291.0,6,Eligible,Long Range,47.37483,-122.19975


#### 6.1 Geospatial (Coordinate Split)

This strategy extracts Latitude and Longitude from the 'Vehicle Location' string column, converting them into separate numeric columns. This is essential for geographical plotting in tools like Tableau.

In [12]:
import re

def extract_coords(point_str):
    if pd.isna(point_str): return None, None
    coords = re.findall(r"[-+]?\d*\.\d+|\d+", str(point_str))
    if len(coords) >= 2: return float(coords[1]), float(coords[0])
    return None, None

# Explicitly create Latitude and Longitude columns and assign values
extracted_data = df['Vehicle Location'].apply(lambda x: pd.Series(extract_coords(x)))
df['Latitude'] = extracted_data[0] # Assuming 0 is Latitude from extract_coords
df['Longitude'] = extracted_data[1] # Assuming 1 is Longitude from extract_coords

print("Columns after Latitude/Longitude extraction:")
print(df.columns.tolist())

Columns after Latitude/Longitude extraction:
['VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year', 'Make', 'Model', 'Electric Vehicle Type', 'Clean Alternative Fuel Vehicle (CAFV) Eligibility', 'Electric Range', 'Legislative District', 'DOL Vehicle ID', 'Vehicle Location', 'Electric Utility', '2020 Census Tract', 'Electric Range Raw', 'Vehicle Age', 'CAFV_Status', 'Range_Category', 'Latitude', 'Longitude']


#### 6.2 Brand Segmentation

This strategy categorizes vehicle manufacturers into 'Luxury' or 'Mass Market' segments. This helps in analyzing market dynamics and consumer preferences across different price points.

In [13]:
luxury_brands = ['TESLA', 'PORSCHE', 'LUCID', 'RIVIAN', 'AUDI', 'BMW', 'MERCEDES-BENZ']
df['Market_Segment'] = df['Make'].apply(lambda x: 'Luxury' if str(x).upper() in luxury_brands else 'Mass Market')
print("'Market_Segment' column created.")
display(df.head())

'Market_Segment' column created.


,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,...,Vehicle Location,Electric Utility,2020 Census Tract,Electric Range Raw,Vehicle Age,CAFV_Status,Range_Category,Latitude,Longitude,Market_Segment
0,1N4AZ0CP6D,King,Kirkland,WA,98034,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,POINT (-122.22901 47.72201),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303302e+10,75.0,13,Eligible,Urban Commuter,47.72201,-122.22901,Mass Market
1,5YJ3E1EC8L,Kitsap,Bainbridge Island,WA,98110,2020,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,POINT (-122.521 47.62759),PUGET SOUND ENERGY INC,5.303509e+10,308.0,6,Eligible,Long Range,47.62759,-122.52100,Luxury
2,5YJ3E1EBXJ,King,Seattle,WA,98144,2018,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,POINT (-122.30866 47.57874),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10,215.0,8,Eligible,Regional Traveler,47.57874,-122.30866,Luxury
3,ZFAFFAC45R,Thurston,Yelm,WA,98597,2024,FIAT,500E,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,...,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10,NaN,2,Unknown/Not Eligible,NaN,46.94239,-122.60735,Mass Market
4,5YJYGDEE3L,King,Kent,WA,98030,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,POINT (-122.19975 47.37483),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303303e+10,291.0,6,Eligible,Long Range,47.37483,-122.19975,Luxury


#### 6.3 Eligibility Traffic Lights

This strategy simplifies the 'Clean Alternative Fuel Vehicle (CAFV) Eligibility' into clearer 'Eligible', 'Not Eligible', or 'Unknown' statuses. This makes it easier to visualize and understand vehicle eligibility.

In [14]:
df['Eligibility_Status'] = df['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].apply(
    lambda x: 'Eligible' if 'Eligible' in str(x) and 'Not' not in str(x) else ('Not Eligible' if 'Not' in str(x) else 'Unknown')
)
print("'Eligibility_Status' column created.")
display(df.head())

'Eligibility_Status' column created.


,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,...,Electric Utility,2020 Census Tract,Electric Range Raw,Vehicle Age,CAFV_Status,Range_Category,Latitude,Longitude,Market_Segment,Eligibility_Status
0,1N4AZ0CP6D,King,Kirkland,WA,98034,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303302e+10,75.0,13,Eligible,Urban Commuter,47.72201,-122.22901,Mass Market,Eligible
1,5YJ3E1EC8L,Kitsap,Bainbridge Island,WA,98110,2020,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,PUGET SOUND ENERGY INC,5.303509e+10,308.0,6,Eligible,Long Range,47.62759,-122.52100,Luxury,Eligible
2,5YJ3E1EBXJ,King,Seattle,WA,98144,2018,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10,215.0,8,Eligible,Regional Traveler,47.57874,-122.30866,Luxury,Eligible
3,ZFAFFAC45R,Thurston,Yelm,WA,98597,2024,FIAT,500E,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,...,PUGET SOUND ENERGY INC,5.306701e+10,NaN,2,Unknown/Not Eligible,NaN,46.94239,-122.60735,Mass Market,Unknown
4,5YJYGDEE3L,King,Kent,WA,98030,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303303e+10,291.0,6,Eligible,Long Range,47.37483,-122.19975,Luxury,Eligible


#### 6.4 Adoption Waves

This strategy categorizes vehicles based on their 'Model Year' into 'Early', 'Growth', or 'Mass Adoption' waves. This helps in understanding the historical progression and market penetration of electric vehicles.

In [15]:
df['Adoption_Wave'] = df['Model Year'].apply(lambda y: 'Early' if y <= 2015 else ('Growth' if y <= 2020 else 'Mass Adoption'))
print("'Adoption_Wave' column created.")
display(df.head())

'Adoption_Wave' column created.


,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,...,2020 Census Tract,Electric Range Raw,Vehicle Age,CAFV_Status,Range_Category,Latitude,Longitude,Market_Segment,Eligibility_Status,Adoption_Wave
0,1N4AZ0CP6D,King,Kirkland,WA,98034,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,5.303302e+10,75.0,13,Eligible,Urban Commuter,47.72201,-122.22901,Mass Market,Eligible,Early
1,5YJ3E1EC8L,Kitsap,Bainbridge Island,WA,98110,2020,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,5.303509e+10,308.0,6,Eligible,Long Range,47.62759,-122.52100,Luxury,Eligible,Growth
2,5YJ3E1EBXJ,King,Seattle,WA,98144,2018,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,5.303301e+10,215.0,8,Eligible,Regional Traveler,47.57874,-122.30866,Luxury,Eligible,Growth
3,ZFAFFAC45R,Thurston,Yelm,WA,98597,2024,FIAT,500E,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,...,5.306701e+10,NaN,2,Unknown/Not Eligible,NaN,46.94239,-122.60735,Mass Market,Unknown,Mass Adoption
4,5YJYGDEE3L,King,Kent,WA,98030,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,5.303303e+10,291.0,6,Eligible,Long Range,47.37483,-122.19975,Luxury,Eligible,Growth


#### 6.5 Utility Complexity (The Infrastructure Load)

This strategy quantifies the 'Utility Complexity Score' based on the number of overlapping jurisdictions mentioned in the 'Electric Utility' column. A higher score might indicate a more complex infrastructure landscape for a vehicle's location.

In [16]:
df['Utility_Complexity_Score'] = df['Electric Utility'].fillna('').apply(
    lambda x: len(x.replace('||', '|').split('|')) if x.strip() != '' else 0
)
print("'Utility_Complexity_Score' column created.")
display(df.head())

'Utility_Complexity_Score' column created.


,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,...,Electric Range Raw,Vehicle Age,CAFV_Status,Range_Category,Latitude,Longitude,Market_Segment,Eligibility_Status,Adoption_Wave,Utility_Complexity_Score
0,1N4AZ0CP6D,King,Kirkland,WA,98034,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,75.0,13,Eligible,Urban Commuter,47.72201,-122.22901,Mass Market,Eligible,Early,2
1,5YJ3E1EC8L,Kitsap,Bainbridge Island,WA,98110,2020,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,308.0,6,Eligible,Long Range,47.62759,-122.52100,Luxury,Eligible,Growth,1
2,5YJ3E1EBXJ,King,Seattle,WA,98144,2018,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,215.0,8,Eligible,Regional Traveler,47.57874,-122.30866,Luxury,Eligible,Growth,2
3,ZFAFFAC45R,Thurston,Yelm,WA,98597,2024,FIAT,500E,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,...,NaN,2,Unknown/Not Eligible,NaN,46.94239,-122.60735,Mass Market,Unknown,Mass Adoption,1
4,5YJYGDEE3L,King,Kent,WA,98030,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,291.0,6,Eligible,Long Range,47.37483,-122.19975,Luxury,Eligible,Growth,2


### Reasoning for Dropping Rows with Missing 'Electric Range Raw'

The 'Electric Range Raw' column contains a significant number of missing (NaN) values, which often correspond to electric vehicles that either have an unknown range or are not pure battery electric vehicles (e.g., Plug-in Hybrid Electric Vehicles where electric range is not a primary metric or was not recorded). For analyses that heavily rely on the electric range, such as understanding range distribution, efficiency, or segmenting vehicles by range capabilities, these missing values can significantly skew results or make certain computations impossible.

Dropping these rows ensures that subsequent analyses, especially those involving `Electric Range Raw` and `Electric Range` (which has 0s converted to NaNs), are based on complete and meaningful data. While it reduces the overall dataset size, it enhances the quality and interpretability of range-specific insights. We will compare the dataset's shape and key statistics before and after this operation to illustrate the impact.

In [ ]:
print("Dataset Comparison: Before vs. After Dropping NaN in 'Electric Range Raw'\n")


initial_shape = df.shape
initial_electric_range_desc = df['Electric Range'].describe()

print(f"Original DataFrame Shape: {initial_shape}")
print("\nDescriptive Statistics for 'Electric Range' (Original):\n")
display(initial_electric_range_desc)

df_after_drop = df.dropna(subset=['Electric Range Raw']).copy()


final_shape = df_after_drop.shape
final_electric_range_desc = df_after_drop['Electric Range'].describe()

print(f"\nDataFrame Shape After Dropping NaN in 'Electric Range Raw': {final_shape}")
print(f"Number of rows dropped: {initial_shape[0] - final_shape[0]}")
print(f"Percentage of data remaining: {round((final_shape[0] / initial_shape[0]) * 100, 2)}%")

print("\nDescriptive Statistics for 'Electric Range' (After Drop):\n")
display(final_electric_range_desc)

df = df_after_drop
print("\nRows with NaN in 'Electric Range Raw' have been dropped from the DataFrame.")

### Dataset Comparison: Before vs. After Dropping NaN in 'Electric Range Raw'

Original DataFrame Shape: (280813, 26)

Descriptive Statistics for 'Electric Range' (Original):



,Electric Range
count,101275.000000
mean,107.462256
std,97.180775
min,1.000000
25%,30.000000
50%,47.000000
75%,215.000000
max,337.000000



DataFrame Shape After Dropping NaN in 'Electric Range Raw': (101275, 26)
Number of rows dropped: 179538
Percentage of data remaining: 36.06%

Descriptive Statistics for 'Electric Range' (After Drop):



,Electric Range
count,101275.000000
mean,107.462256
std,97.180775
min,1.000000
25%,30.000000
50%,47.000000
75%,215.000000
max,337.000000



Rows with NaN in 'Electric Range Raw' have been dropped from the DataFrame.


### 7.1 Final Cleanup for Export

This step performs a final cleanup by removing rows where Latitude or Longitude could not be extracted (as these are critical for mapping visualizations) and then exports the newly cleaned and enriched dataset to a CSV file ready for Tableau.

In [18]:
df = df.dropna(subset=['Latitude', 'Longitude']) # Remove rows without location for Tableau
df.to_csv('/content/EV_Population_Cleaned.csv', index=False)
print("Cleaning Complete. New features created for Tableau.")
display(df.head())

Cleaning Complete. New features created for Tableau.


,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,...,Electric Range Raw,Vehicle Age,CAFV_Status,Range_Category,Latitude,Longitude,Market_Segment,Eligibility_Status,Adoption_Wave,Utility_Complexity_Score
0,1N4AZ0CP6D,King,Kirkland,WA,98034,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,75.0,13,Eligible,Urban Commuter,47.72201,-122.22901,Mass Market,Eligible,Early,2
1,5YJ3E1EC8L,Kitsap,Bainbridge Island,WA,98110,2020,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,308.0,6,Eligible,Long Range,47.62759,-122.52100,Luxury,Eligible,Growth,1
2,5YJ3E1EBXJ,King,Seattle,WA,98144,2018,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,215.0,8,Eligible,Regional Traveler,47.57874,-122.30866,Luxury,Eligible,Growth,2
4,5YJYGDEE3L,King,Kent,WA,98030,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,...,291.0,6,Eligible,Long Range,47.37483,-122.19975,Luxury,Eligible,Growth,2
5,WBY1Z4C50F,Snohomish,Mukilteo,WA,98275,2015,BMW,I3,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,...,72.0,11,Eligible,Urban Commuter,47.89908,-122.29196,Luxury,Eligible,Early,1


In [19]:
from google.colab import files
files.download("EV_Population_Cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Data Cleaning and Feature Engineering Report for Electric Vehicle Population Dataset

This report details the comprehensive data cleaning, standardization, and feature engineering steps applied to the Electric Vehicle Population dataset. The goal of these operations was to prepare the data for further analysis and visualization, specifically for tools like Tableau.

### 1. Initial Setup and Data Loading
-   **Libraries Imported**: `pandas` and `numpy` were imported for data manipulation.
-   **Dataset Loading**: The dataset `Electric_Vehicle_Population_Data.csv` was loaded into a pandas DataFrame named `df`. A `DtypeWarning` was noted for `Columns (11)` due to mixed types, handled implicitly by pandas.

### 2. Data Cleaning & Standardization
-   **Missing Geographic Data**: Missing values in the following geographic columns were filled with the string 'Unknown':
    -   `County`
    -   `City`
    -   `Postal Code`
    -   `Legislative District`
    -   `Vehicle Location`
-   **Postal Code Standardization**: The `Postal Code` column was standardized by converting it to a string type and removing any trailing '.0' (e.g., '98034.0' became '98034').

### 3. Handling 'Electric Range'
-   **'Electric Range Raw' Creation**: A new column `Electric Range Raw` was created as a copy of the original `Electric Range` column to preserve the raw values.
-   **Zero Value Treatment**: All zero values in the `Electric Range` column were replaced with `np.nan`. This was done to ensure that statistical analyses on electric range would not be skewed by vehicles reported with a '0' range, which often indicates an unknown or unrecorded value rather than a literal zero range.

### 4. Feature Engineering (KPI Preparation)
-   **'Vehicle Age' Calculation**: A `Vehicle Age` column was created by subtracting the `Model Year` from a `current_year` set to `2026`. This provides a simple age metric for each vehicle.
-   **'CAFV_Status' Simplification**: The `Clean Alternative Fuel Vehicle (CAFV) Eligibility` column was simplified into a new column `CAFV_Status`. Vehicles containing 'Eligible' are classified as 'Eligible', otherwise 'Unknown/Not Eligible'.

### 5. Advanced Feature Engineering for Tableau Visualization
This section focused on creating new features that are particularly useful for detailed visualizations and segmentation in analytical tools.

-   **'Range_Category' Creation**: A categorical column `Range_Category` was derived from `Electric Range Raw` using predefined bins and labels:
    -   `[-1, 0]`: 'Unknown/Not Researched'
    -   `(0, 100]`: 'Urban Commuter'
    -   `(100, 250]`: 'Regional Traveler'
    -   `(250, 1000]`: 'Long Range'

-   **Geospatial (Coordinate Split)**:
    -   A helper function `extract_coords` was defined to parse `Latitude` and `Longitude` from the `Vehicle Location` string column using regular expressions.
    -   Two new numeric columns, `Latitude` and `Longitude`, were created using this function, enabling geographical plotting.

-   **Brand Segmentation (`Market_Segment`)**:
    -   A list of `luxury_brands` was defined (`TESLA`, `PORSCHE`, `LUCID`, `RIVIAN`, `AUDI`, `BMW`, `MERCEDES-BENZ`).
    -   A `Market_Segment` column was created, classifying vehicles as 'Luxury' if their `Make` is in the `luxury_brands` list, and 'Mass Market' otherwise.

-   **Eligibility Traffic Lights (`Eligibility_Status`)**:
    -   A `Eligibility_Status` column was created to simplify the `Clean Alternative Fuel Vehicle (CAFV) Eligibility` into clearer categories: 'Eligible', 'Not Eligible', or 'Unknown'.

-   **Adoption Waves (`Adoption_Wave`)**:
    -   The `Model Year` was used to categorize vehicles into `Adoption_Wave`:
        -   `Model Year <= 2015`: 'Early'
        -   `2015 < Model Year <= 2020`: 'Growth'
        -   `Model Year > 2020`: 'Mass Adoption'

-   **Utility Complexity (`Utility_Complexity_Score`)**:
    -   A `Utility_Complexity_Score` was calculated based on the number of distinct utility jurisdictions listed in the `Electric Utility` column. The score is derived by splitting the string by '|' and counting the resulting segments.

### 6. Handling Missing 'Electric Range Raw' Values
-   **Decision Rationale**: A significant number of rows (179,538 out of 280,813) had `NaN` values in `Electric Range Raw`. For analyses heavily reliant on electric range, these missing values would skew results. Therefore, a decision was made to drop these rows to ensure data quality and interpretability for range-specific insights.
-   **Impact**: The DataFrame shape was reduced from `(280813, 26)` to `(101275, 26)`, retaining approximately `36.06%` of the original data. This ensures that subsequent analyses only use records with a defined electric range.

### 7. Final Cleanup and Export
-   **Geospatial Data Filtering**: Rows with missing `Latitude` or `Longitude` (which could not be extracted or were originally null) were removed. This is crucial for accurate mapping visualizations.
-   **Final Export**: The thoroughly cleaned and enriched DataFrame was exported to a CSV file named `EV_Population_Cleaned.csv` in the `/content/` directory, ready for use in external visualization tools like Tableau.